examples/gno.land/p/demo/gnorkle/gnorkle/instance.gno

```diff
@@ -0,0 +1,197 @@
+package gnorkle
+
+import (
+	"std"
+	"strings"
+
+	"gno.land/p/demo/avl"
+	"gno.land/p/demo/gnorkle/agent"
+	"gno.land/p/demo/gnorkle/feed"
+	"gno.land/p/demo/gnorkle/message"
+)
+
+// Instance is a single instance of an oracle.
+type Instance struct {
+	feeds     *avl.Tree
+	whitelist agent.Whitelist
+}
+
+// NewInstance creates a new instance of an oracle.
+func NewInstance() *Instance {
+	return &Instance{
+		feeds: avl.NewTree(),
+	}
+}
+
+func assertValidID(id string) {
+	if len(id) == 0 {
+		panic("feed ids cannot be empty")
+	}
+
+	if strings.Contains(id, ",") {
+		panic("feed ids cannot contain commas")
+	}
+}
+
+func (i *Instance) assertFeedDoesNotExist(id string) {
+	if i.feeds.Has(id) {
+		panic("feed already exists")
+	}
+}
+
+// AddFeeds adds feeds to the instance with empty whitelists.
+func (i *Instance) AddFeeds(feeds ...Feed) {
+	for _, feed := range feeds {
+		assertValidID(feed.ID())
+		i.assertFeedDoesNotExist(feed.ID())
+		i.feeds.Set(
+			feed.ID(),
+			FeedWithWhitelist{
+				Whitelist: new(agent.Whitelist),
+				Feed:      feed,
+			},
+		)
+	}
+}
+
+// AddFeedsWithWhitelists adds feeds to the instance with the given whitelists.
+func (i *Instance) AddFeedsWithWhitelists(feeds ...FeedWithWhitelist) {
+	for _, feed := range feeds {
+		i.assertFeedDoesNotExist(feed.ID())
+		assertValidID(feed.ID())
+		i.feeds.Set(
+			feed.ID(),
+			FeedWithWhitelist{
+				Whitelist: feed.Whitelist,
+				Feed:      feed,
+			},
+		)
+	}
+}
+
+// RemoveFeed removes a feed from the instance.
+func (i *Instance) RemoveFeed(id string) {
+	i.feeds.Remove(id)
+}
+
+// PostMessageHandler is a type that allows for post-processing of feed state after a feed
+// ingests a message from an agent.
+type PostMessageHandler interface {
+	Handle(i *Instance, funcType message.FuncType, feed Feed)
+}
+
```

Comment:
a general comment, which I think is important here but also elsewhere

I see that there's many `panics` called throughout this package in place of error handling. However, I think what's written in [Effective Gno](https://docs.gno.land/concepts/effective-gno#embrace-panic) is important, and in the package itself we should try using errors as we do in Go rather than panicking on any errors. Of course some cases make sense (generally, the same cases where we panic in Go, or `Must*` / `Assert*` functions), but I think it makes sense to change many functions here to be more "Go-like".

I'm commenting here because I first noticed this when seeing `Handle` doesn't return an error value. I think it makes sense if it does, though it's OK if the realm instead panics. As I see it, the realm should decide when to panic, as it is the "entity" that deals with the transaction itself, whereas the package should not.

Reply:
Good call; makes sense to make.  I'll make the changes

Reply:
Done